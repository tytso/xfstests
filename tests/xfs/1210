#! /bin/bash
# SPDX-License-Identifier: GPL-2.0-or-later
# Copyright (c) 2021 Oracle.  All Rights Reserved.
#
# FS QA Test No. 1210
#
# Check for proper operation of the atomicswap log incompat feature flag
# when crashes happen.  That is to say:
# - It gets turned on after we try to use swapext;
# - It is still set after we crash while using swapext and gets cleared after
#   recovery;
# - It is still set after we crash while using swapext and gets cleared after
#   ro recovery.

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1    # failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/inject

# real QA test starts here
_supported_fs xfs
_require_xfs_scratch_atomicswap
_require_xfs_io_error_injection "bmap_finish_one"

rm -f $seqres.full

has_atomicswap() {
	local val="$(_scratch_xfs_db -f -c 'sb 0' -c 'print features_log_incompat' | cut -d ' ' -f 3)"
	echo "current log incompat: $val" >> $seqres.full
	# XFS_SB_FEAT_INCOMPAT_LOG_ATOMIC_SWAP == 1
	test "$(( $val & 0x1 ))" -eq 1 && return 0
	return 1
}

echo "Format filesystem" | tee -a $seqres.full
_scratch_mkfs -m reflink=1,rmapbt=1 > $seqres.full
has_atomicswap && echo "atomic swap feature should not be enabled"

echo "Make sure atomicswap gets turned on during usage." | tee -a $seqres.full
_scratch_mount
blksz=65536
_pwrite_byte 0x59 0 $blksz $SCRATCH_MNT/a >> $seqres.full
_pwrite_byte 0x60 0 $blksz $SCRATCH_MNT/b >> $seqres.full
md5sum $SCRATCH_MNT/a $SCRATCH_MNT/b | _filter_scratch

$XFS_IO_PROG -c "swapext -a $SCRATCH_MNT/b" $SCRATCH_MNT/a
has_atomicswap || echo "atomic swap feature should be enabled"

echo "Check swap results" | tee -a $seqres.full
_scratch_cycle_mount
md5sum $SCRATCH_MNT/a $SCRATCH_MNT/b | _filter_scratch

echo "Check set after a shutdown" | tee -a $seqres.full
$XFS_IO_PROG -x -c 'inject bmap_finish_one' \
	-c "swapext -a $SCRATCH_MNT/b" $SCRATCH_MNT/a
_scratch_unmount
has_atomicswap || echo "atomic swap feature should be enabled"

echo "Check clear after recovery" | tee -a $seqres.full
_scratch_mount
md5sum $SCRATCH_MNT/a $SCRATCH_MNT/b | _filter_scratch
_scratch_unmount
has_atomicswap && echo "atomic swap feature should not be enabled"
_scratch_mount

echo "Check set after a second shutdown" | tee -a $seqres.full
$XFS_IO_PROG -x -c 'inject bmap_finish_one' \
	-c "swapext -a $SCRATCH_MNT/b" $SCRATCH_MNT/a
_scratch_unmount
has_atomicswap || echo "atomic swap feature should be enabled"

echo "Check clear after readonly recovery" | tee -a $seqres.full
_scratch_mount -o ro
md5sum $SCRATCH_MNT/a $SCRATCH_MNT/b | _filter_scratch
_scratch_unmount
has_atomicswap && echo "atomic swap feature should not be enabled"
_scratch_mount

status=0
exit
