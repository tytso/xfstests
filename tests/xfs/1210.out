QA output created by 1210
Format filesystem
Make sure atomicswap gets turned on during usage.
1fed14479747f703642f26a1eba137dd  SCRATCH_MNT/a
794faddd00bd8ef1ccfdf4fea4381fbe  SCRATCH_MNT/b
Check swap results
794faddd00bd8ef1ccfdf4fea4381fbe  SCRATCH_MNT/a
1fed14479747f703642f26a1eba137dd  SCRATCH_MNT/b
Check set after a shutdown
swapext: Input/output error
Check clear after recovery
1fed14479747f703642f26a1eba137dd  SCRATCH_MNT/a
794faddd00bd8ef1ccfdf4fea4381fbe  SCRATCH_MNT/b
Check set after a second shutdown
swapext: Input/output error
Check clear after readonly recovery
794faddd00bd8ef1ccfdf4fea4381fbe  SCRATCH_MNT/a
1fed14479747f703642f26a1eba137dd  SCRATCH_MNT/b
